gnd/src/main/java/com/google/android/gnd/workers/TileDownloadWorker.java

```diff
@@ -42,106 +45,106 @@
    */
-  private Result downloadTileFile(Tile tile, Optional<HashMap<String, String>> requestProperties) {
-    try {
-      URL url = new URL(tile.getUrl());
-      HttpURLConnection connection = (HttpURLConnection) url.openConnection();
-
-      if (requestProperties.isPresent()) {
-        for (Map.Entry<String, String> property : requestProperties.get().entrySet()) {
-          connection.setRequestProperty(property.getKey(), property.getValue());
-        }
-      }
-
-      connection.connect();
-
-      InputStream is = connection.getInputStream();
-      FileOutputStream fos = context.openFileOutput(tile.getPath(), Context.MODE_PRIVATE);
-      byte[] byteChunk = new byte[BUFFER_SIZE];
-      int n;
-
-      while ((n = is.read(byteChunk)) > 0) {
-        fos.write(byteChunk, 0, n);
+  private void downloadTileFile(Tile tile, Map<String, String> requestProperties)
+      throws IOException {
+    URL url = new URL(tile.getUrl());
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+
+    if (!requestProperties.isEmpty()) {
+      for (Map.Entry<String, String> property : requestProperties.entrySet()) {
+        connection.setRequestProperty(property.getKey(), property.getValue());
       }
+    }
 
-      is.close();
-      fos.close();
-
-      localDataStore
-          .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.DOWNLOADED).build())
-          .blockingAwait();
-
-      return Result.success();
-
-    } catch (IOException e) {
-      Log.d(TAG, "Failed to download and write file.", e);
+    connection.connect();
 
-      localDataStore
-          .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.FAILED).build())
-          .blockingAwait();
+    InputStream is = connection.getInputStream();
+    FileOutputStream fos = context.openFileOutput(tile.getPath(), Context.MODE_PRIVATE);
+    byte[] byteChunk = new byte[BUFFER_SIZE];
+    int n;
 
-      return Result.failure();
+    while ((n = is.read(byteChunk)) > 0) {
+      fos.write(byteChunk, 0, n);
     }
-  }
-
-  /** Update a tile's state in the database and initiate a download of the tile source file. */
-  private Result downloadTile(Tile tile) {
-    localDataStore
-        .insertOrUpdateTile(tile.toBuilder().setState(Tile.State.IN_PROGRESS).build())
-        .blockingAwait();
 
-    return downloadTileFile(tile, Optional.empty());
+    is.close();
+    fos.close();
   }
 
-  /** Resumes downloading the source for {@param tile} marked as {@code Tile.State.IN_PROGRESS}. */
-  private Result resumeTileDownload(Tile tile) {
-    File existingTileFile = new File(context.getFilesDir(), tile.getPath());
+  /** Update a tile's state in the database and initiate a download of the tile source file. */
+  private Completable downloadTile(Tile tile) {
     HashMap<String, String> requestProperties = new HashMap<>();
 
-    requestProperties.put("Range", existingTileFile.length() + "-");
+    if (tile.getState() == State.IN_PROGRESS) {
```

Comment:
Please add a comment explaining why we do this (to result downloads)?

Reply:
The range request is supposed to let us [resume the file download from where it stopped](https://developer.mozilla.org/en-US/docs/Web/HTTP/Range_requests), but tbh I haven't confirmed it works in practice, and I don't really know much about HTTP range requests at all.

It might be safer to just start the entire file download again when we resume something in progress and overwrite anything that was already on disk. Even though we're technically doing extra work in that case.

Reply:
Hmm.. perhaps.. we can leave it for now provided we test it out.

Reply:
sg!

Reply:
Fixed some problems w/ this and added explanatory comments in de5953f